pkg/query-service/rules/manager.go

```diff
@@ -593,3 +603,108 @@ func (m *Manager) GetRule(id string) (*GettableRule, error) {
 	r.Id = fmt.Sprintf("%d", s.Id)
 	return r, nil
 }
+
+// syncRuleStateWithTask ensures that the state of a stored rule matches
+// the task state. For example - if a stored rule is disabled, then
+// there is no task running against it.
+func (m *Manager) syncRuleStateWithTask(taskName string, rule *PostableRule) error {
+
+	if rule.Disabled {
+		// check if rule has any task running
+		if _, ok := m.tasks[taskName]; ok {
+			// delete task from memory
+			if err := m.deleteTask(taskName); err != nil {
+				return err
+			}
+		}
+	} else {
+		// check if rule has a task running
+		if _, ok := m.tasks[taskName]; !ok {
+			// rule has not task, start one
+			if err := m.addTask(rule, taskName); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// PatchRule supports attribute level changes to the rule definition unlike
+// EditRule, which updates entire rule definition in the DB.
+// the process:
+//  - get the latest rule from db
+//  - over write the patch attributes received in input (ruleStr)
+//  - re-deploy or undeploy task as necessary
+//  - update the patched rule in the DB
+func (m *Manager) PatchRule(ruleStr string, ruleId string) (*GettableRule, error) {
+
+	if ruleId == "" {
+		return nil, fmt.Errorf("id is mandatory for patching rule")
+	}
+
+	taskName := prepareTaskName(ruleId)
+
+	// retrieve rule from DB
+	storedJSON, err := m.ruleDB.GetStoredRule(ruleId)
+	if err != nil {
+		zap.S().Errorf("msg:", "failed to get stored rule with given id", "\t error:", err)
+		return nil, err
+	}
+
+	// storedRule holds the current stored rule from DB
+	storedRule := PostableRule{}
+	if err := json.Unmarshal([]byte(storedJSON.Data), &storedRule); err != nil {
+		zap.S().Errorf("msg:", "failed to get unmarshal stored rule with given id", "\t error:", err)
+		return nil, err
+	}
+
+	// parsed Rule is combo of stored rule and patch received in request
+	patchedRule, errs := parseIntoRule(storedRule, []byte(ruleStr), "json")
+	if len(errs) > 0 {
+		zap.S().Errorf("failed to parse rules:", errs)
+		// just one rule is being parsed so expect just one error
+		return nil, errs[0]
+	}
+
+	if err := m.syncRuleStateWithTask(taskName, patchedRule); err != nil {
+		zap.S().Errorf("failed to sync stored rule state with the task")
+		return nil, err
+	}
+
+	// prepare rule json to write to update db
+	patchedRuleBytes, err := json.Marshal(patchedRule)
+	if err != nil {
+		return nil, err
+	}
+
+	// write updated rule to db
+	if _, _, err = m.ruleDB.EditRuleTx(string(patchedRuleBytes), ruleId); err != nil {
+		// write failed, rollback task state
+
+		// restore task state from the stored rule
+		if err := m.syncRuleStateWithTask(taskName, &storedRule); err != nil {
+			zap.S().Errorf("msg: ", "failed to restore rule after patch failure", "\t error:", err)
+		}
+
+		return nil, err
+	}
+
+	// prepare http response
+	response := GettableRule{}
+	if err := json.Unmarshal(patchedRuleBytes, &response); err != nil { // Parse []byte to go struct pointer
```

Comment:
I find it uneasy to number marshal/unmarshal in patch code. I wish we could just update the entry instead of back and forth b/w marshal and unmarshal.

Reply:
You are right. It's a mess.  Let me re-think on this.